Panic due to missing init container

**Describe the bug**

PR #562 introduced a bug where a panic occurs if a version of Argo CD is installed on the cluster by an older version of the operator that didn't have the CMP 2.0 support included.

```
panic: runtime error: index out of range [0] with length 0

goroutine 505 [running]:
github.com/argoproj-labs/argocd-operator/controllers/argocd.(*ReconcileArgoCD).reconcileRepoDeployment(0xc000ac2d20, 0xc0007b5800)
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/controllers/argocd/deployment.go:1060 +0x2773
github.com/argoproj-labs/argocd-operator/controllers/argocd.(*ReconcileArgoCD).reconcileDeployments(0x27e1e80, 0xc0003e1640)
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/controllers/argocd/deployment.go:347 +0x6c
github.com/argoproj-labs/argocd-operator/controllers/argocd.(*ReconcileArgoCD).reconcileResources(0xc000ac2d20, 0xc0007b5800)
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/controllers/argocd/util.go:755 +0x2a7
github.com/argoproj-labs/argocd-operator/controllers/argocd.(*ReconcileArgoCD).Reconcile(0xc000ac2d20, {0x27c5cf8, 0xc0008883c0}, {{{0xc0005ff0b9, 0x24aa140}, {0xc000b06930, 0x30}}})
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/controllers/argocd/argocd_controller.go:122 +0x42e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc00023c580, {0x27c5cf8, 0xc0008882d0}, {{{0xc0005ff0b9, 0x24aa140}, {0xc000b06930, 0x1014f54}}})
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114 +0x26f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00023c580, {0x27c5c50, 0xc0005cd6c0}, {0x23cffe0, 0xc000b40b00})
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311 +0x33e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00023c580, {0x27c5c50, 0xc0005cd6c0})
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/Users/aveerama/go/src/github.com/iam-veeramalla/argocd-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:223 +0x357```

--- a/controllers/argocd/deployment.go
+++ b/controllers/argocd/deployment.go
@@ -1060,24 +1060,6 @@ func (r *ReconcileArgoCD) reconcileRepoDeployment(cr *argoprojv1a1.ArgoCD) error
 			existing.Spec.Template.Spec.InitContainers = deploy.Spec.Template.Spec.InitContainers
 			changed = true
 		}
-		if !reflect.DeepEqual(deploy.Spec.Template.Spec.InitContainers[0].VolumeMounts,
-			existing.Spec.Template.Spec.InitContainers[0].VolumeMounts) {
-			existing.Spec.Template.Spec.InitContainers[0].VolumeMounts = deploy.Spec.Template.Spec.InitContainers[0].VolumeMounts
-			changed = true
-		}
-		if !reflect.DeepEqual(deploy.Spec.Template.Spec.InitContainers[0].Env,
-			existing.Spec.Template.Spec.InitContainers[0].Env) {
-			existing.Spec.Template.Spec.InitContainers[0].Env = deploy.Spec.Template.Spec.InitContainers[0].Env
-			changed = true
-		}
-		if !reflect.DeepEqual(deploy.Spec.Template.Spec.InitContainers[0].Resources, existing.Spec.Template.Spec.InitContainers[0].Resources) {
-			existing.Spec.Template.Spec.InitContainers[0].Resources = deploy.Spec.Template.Spec.InitContainers[0].Resources
-			changed = true
-		}
-		if !reflect.DeepEqual(deploy.Spec.Template.Spec.InitContainers[0].Command, existing.Spec.Template.Spec.InitContainers[0].Command) {
-			existing.Spec.Template.Spec.InitContainers[0].Command = deploy.Spec.Template.Spec.InitContainers[0].Command
-			changed = true
-		}
 		if !reflect.DeepEqual(deploy.Spec.Replicas, existing.Spec.Replicas) {
 			existing.Spec.Replicas = deploy.Spec.Replicas
 			changed = true

--- a/controllers/argocd/deployment.go
+++ b/controllers/argocd/deployment.go
@@ -1056,6 +1056,10 @@ func (r *ReconcileArgoCD) reconcileRepoDeployment(cr *argoprojv1a1.ArgoCD) error
 				deploy.Spec.Template.Spec.Containers[1:]...)
 			changed = true
 		}
+		if !reflect.DeepEqual(deploy.Spec.Template.Spec.InitContainers, existing.Spec.Template.Spec.InitContainers) {
+			existing.Spec.Template.Spec.InitContainers = deploy.Spec.Template.Spec.InitContainers
+			changed = true
+		}
 		if !reflect.DeepEqual(deploy.Spec.Template.Spec.InitContainers[0].VolumeMounts,
 			existing.Spec.Template.Spec.InitContainers[0].VolumeMounts) {
 			existing.Spec.Template.Spec.InitContainers[0].VolumeMounts = deploy.Spec.Template.Spec.InitContainers[0].VolumeMounts
@@ -1072,12 +1076,6 @@ func (r *ReconcileArgoCD) reconcileRepoDeployment(cr *argoprojv1a1.ArgoCD) error
 		}
 		if !reflect.DeepEqual(deploy.Spec.Template.Spec.InitContainers[0].Command, existing.Spec.Template.Spec.InitContainers[0].Command) {
 			existing.Spec.Template.Spec.InitContainers[0].Command = deploy.Spec.Template.Spec.InitContainers[0].Command
-			changed = true
-		}
-		if !reflect.DeepEqual(deploy.Spec.Template.Spec.InitContainers[1:],
-			existing.Spec.Template.Spec.InitContainers[1:]) {
-			existing.Spec.Template.Spec.InitContainers = append(existing.Spec.Template.Spec.InitContainers[0:1],
-				deploy.Spec.Template.Spec.InitContainers[1:]...)
 			changed = true
 		}
 		if !reflect.DeepEqual(deploy.Spec.Replicas, existing.Spec.Replicas) {

--- a/controllers/argocd/deployment_test.go
+++ b/controllers/argocd/deployment_test.go
@@ -425,6 +425,83 @@ func TestReconcileArgoCD_reconcileRepoDeployment_initContainers(t *testing.T) {
 	assert.Equal(t, deployment.Spec.Template.Spec.InitContainers[1].Name, "test-init-container")
 }
 
+func TestReconcileArgoCD_reconcileRepoDeployment_missingInitContainers(t *testing.T) {
+	logf.SetLogger(ZapLogger(true))
+	a := makeTestArgoCD()
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "argocd-repo-server",
+			Namespace: testNamespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Command: []string{"testing"},
+							Image:   "test-image",
+						},
+					},
+					InitContainers: []corev1.Container{},
+				},
+			},
+		},
+	}
+	r := makeTestReconciler(t, a, d)
+
+	err := r.reconcileRepoDeployment(a)
+	assert.NoError(t, err)
+	deployment := &appsv1.Deployment{}
+	err = r.Client.Get(context.TODO(), types.NamespacedName{
+		Name:      "argocd-repo-server",
+		Namespace: testNamespace,
+	}, deployment)
+	assert.NoError(t, err)
+	assert.Len(t, deployment.Spec.Template.Spec.InitContainers, 1)
+	assert.Equal(t, deployment.Spec.Template.Spec.InitContainers[0].Name, "copyutil")
+}
+func TestReconcileArgoCD_reconcileRepoDeployment_unexpectedInitContainer(t *testing.T) {
+	logf.SetLogger(ZapLogger(true))
+	a := makeTestArgoCD()
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "argocd-repo-server",
+			Namespace: testNamespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Command: []string{"testing"},
+							Image:   "test-image",
+						},
+					},
+					InitContainers: []corev1.Container{
+						{
+							Name:    "unknown",
+							Command: []string{"testing-ic"},
+							Image:   "test-image-ic",
+						},
+					},
+				},
+			},
+		},
+	}
+	r := makeTestReconciler(t, a, d)
+
+	err := r.reconcileRepoDeployment(a)
+	assert.NoError(t, err)
+	deployment := &appsv1.Deployment{}
+	err = r.Client.Get(context.TODO(), types.NamespacedName{
+		Name:      "argocd-repo-server",
+		Namespace: testNamespace,
+	}, deployment)
+	assert.NoError(t, err)
+	assert.Len(t, deployment.Spec.Template.Spec.InitContainers, 1)
+	assert.Equal(t, deployment.Spec.Template.Spec.InitContainers[0].Name, "copyutil")
+}
+
 func TestReconcileArgoCD_reconcileRepoDeployment_command(t *testing.T) {
 	logf.SetLogger(ZapLogger(true))
 	a := makeTestArgoCD()

